chore: Adds optional package-lock-filename parameter to test-nodejs*

AndrewSamaha opened the pull request:
[SMS-1606](https://goodwaygroup.atlassian.net/browse/SMS-1606?atlOrigin=eyJpIjoiMDcyMTFhZTZjNGZmNDdlYzljZmI4MTc0OTQ1MzkwOGQiLCJwIjoiaiJ9)

We need to be able to specify the package-lock filename used as a cache key for CircleCI in order to use package managers like pnpm. [This PR](https://github.com/GoodwayGroup/goes/pull/13/files) in our goes repo adds this parameter. The current PR (yes, this one you're reading now) accomplishes that task.

There's a question in my mind about whether the parameter should be still surrounded by quotes. [This post](https://discuss.circleci.com/t/setting-cache-directory-for-save-cache-conditionally/49030/4) on discuss.circleci.com seems to suggest that doing so is correct. However, perplexity.ai makes a compelling case for leaving out the quotes:

```
Based on the information you've provided, when parameterizing the filename in a CircleCI orb file command,
you should not wrap the parameter in quotes. The correct way to use the parameter in this context is:
Here's why:
CircleCI uses double curly braces {{ }} for variable substitution and command execution.
The << >> syntax is used for parameter substitution in orbs.
The checksum function is expecting a string argument, but the parameter substitution will provide that string.
By using this format, CircleCI will:
First, substitute the parameter value for << parameters.package-lock-filename >>.
Then, execute the checksum function on the resulting filename.
```

[SMS-1606]: https://goodwaygroup.atlassian.net/browse/SMS-1606?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

# Testing
This workflow seems to let one validate the pattern locally:
```
❯ circleci orb pack src/ > orbtest.yml
❯ circleci orb validate ./orbtest.yml
Orb at `./orbtest.yml` is valid.
❯ cat orbtest.yml | grep package-lock-filename
            package-lock-filename:
                    - << parameters.cache-key >>-{{ checksum << parameters.package-lock-filename >> }}
                key: << parameters.cache-key >>-{{ checksum << parameters.package-lock-filename >> }}
            package-lock-filename:
                package-lock-filename: << parameters.package-lock-filename >>
            package-lock-filename:
                package-lock-filename: << parameters.package-lock-filename >>
```

# Callouts
I did NOT add the parameter to the apollo-service-task step as AFAIK, we don't have plans to add pnpm to any apollo projects and I wanted to minimize the surface area in case this change broke something.

AndrewSamaha committed:
chore: Adds optional package-lock-filename parameter to test-nodejs* jobs

nathanialg commented:
@AndrewSamaha If you haven't already, could you download the CircleCI CLI tool and validate the changes?
https://circleci.com/docs/how-to-use-the-circleci-local-cli/#validate-a-circleci-config
https://circleci.com/docs/how-to-use-the-circleci-local-cli/#validate-an-orb-in-your-configuration-file

AndrewSamaha commented:
> @AndrewSamaha If you haven't already, could you download the CircleCI CLI tool and validate the changes? https://circleci.com/docs/how-to-use-the-circleci-local-cli/#validate-a-circleci-config https://circleci.com/docs/how-to-use-the-circleci-local-cli/#validate-an-orb-in-your-configuration-file

@nathanialg -- I can't seem to get the tool to validate even the master branch orb correctly. Am I doing it right?
```
❯ circleci config validate src/@orb.yml
Error: config compilation contains errors: config compilation contains errors:
        - There are no workflows or build jobs in the config.
❯ git checkout master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
❯ circleci config validate src/@orb.yml
Error: config compilation contains errors: config compilation contains errors:
        - There are no workflows or build jobs in the config.
❯ circleci config validate ./src/@orb.yml
Error: config compilation contains errors: config compilation contains errors:
        - There are no workflows or build jobs in the config.
❯ git checkout feat/SMS-1606
Switched to branch 'feat/SMS-1606'
Your branch is up to date with 'origin/feat/SMS-1606'.
❯ circleci config validate ./src/@orb.yml
Error: config compilation contains errors: config compilation contains errors:
        - There are no workflows or build jobs in the config.
```

AndrewSamaha commented:
@nathanialg , actually, I think I got it. Here's the workflow:
```
❯ circleci config pack . > tmp.yml
❯ git status
On branch feat/SMS-1606
Your branch is up to date with 'origin/feat/SMS-1606'.

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        tmp.yml

nothing added to commit but untracked files present (use "git add" to track)
❯ circleci config validate ./tmp.yml
Error: The config is using a deprecated Linux VM image (ubuntu-1604:201903-01). Please see https://circleci.com/blog/ubuntu-14-16-image-deprecation/. This error can be ignored by using the '--ignore-deprecated-images' flag.
❯ circleci config validate ./tmp.yml --ignore-deprecated-images
Config file at ./tmp.yml is valid.
```

AndrewSamaha committed:
fix: Removes quotes from new parameter

AndrewSamaha commented:
And, another actually, closer inspection of the resulting tmp.yml shows that it does NOT contain the logic being tested. So, I don't think the workflow above is conclusive validation.

AndrewSamaha commented:
This workflow seems to finally validate the logic:
```
❯ circleci orb pack src/ > orbtest.yml
❯ circleci orb validate ./orbtest.yml
Orb at `./orbtest.yml` is valid.
❯ cat orbtest.yml | grep package-lock-filename
            package-lock-filename:
                    - << parameters.cache-key >>-{{ checksum << parameters.package-lock-filename >> }}
                key: << parameters.cache-key >>-{{ checksum << parameters.package-lock-filename >> }}
            package-lock-filename:
                package-lock-filename: << parameters.package-lock-filename >>
            package-lock-filename:
                package-lock-filename: << parameters.package-lock-filename >>
```

nathanialg reviewed (approved):
👍
